csharp/src/Apache.Arrow/Ipc/ArrowStreamReaderImplementation.cs

```diff
@@ -51,69 +54,101 @@ public override async Task<RecordBatch> ReadNextRecordBatchAsync(CancellationTok
 
         public override RecordBatch ReadNextRecordBatch()
         {
-            throw new NotImplementedException();
+            return ReadRecordBatch();
         }
 
         protected async Task<RecordBatch> ReadRecordBatchAsync(CancellationToken cancellationToken = default)
         {
             await ReadSchemaAsync().ConfigureAwait(false);
 
-            var bytesRead = 0;
+            int bytesRead = 0;
 
-            byte[] lengthBuffer = null;
-            byte[] messageBuff = null;
-            byte[] bodyBuff = null;
-
-            try
+            int messageLength = 0;
+            await Buffers.RentReturnAsync(4, async (lengthBuffer) =>
             {
                 // Get Length of record batch for message header.
+                bytesRead += await BaseStream.ReadAsync(lengthBuffer.AsMemory(0, 4), cancellationToken)
+                    .ConfigureAwait(false);
+
+                if (bytesRead == 4)
+                {
+                    messageLength = BitConverter.ToInt32(lengthBuffer, 0);
+                }
+            }).ConfigureAwait(false);
+
+            if (messageLength == 0)
+            {
+                // reached end
+                return null;
+            }
+
+            RecordBatch result = null;
+            await Buffers.RentReturnAsync(messageLength, async (messageBuff) =>
+            {
+                bytesRead += await BaseStream.ReadAsync(messageBuff.AsMemory(0, messageLength), cancellationToken)
+                    .ConfigureAwait(false);
+                Flatbuf.Message message = Flatbuf.Message.GetRootAsMessage(new FlatBuffers.ByteBuffer(messageBuff));
 
-                lengthBuffer = Buffers.Rent(4);
-                bytesRead += await BaseStream.ReadAsync(lengthBuffer, 0, 4, cancellationToken)
+                Memory<byte> bodyBuff = _memoryPool.Allocate((int)message.BodyLength);
```

Comment:
Where and when do these buffers get returned to the pool?

Reply:
I have changed this to return an `IMemoryOwner<byte>` which is then wrapped by `IDisposable` objects. When those objects (`RecordBatch`, `ArrowBuffer`) get disposed, they dispose the `IMemoryOwner`.